### PR TITLE
Add "Picture in picture" plugin

### PR DESCRIPTION
**Pull request by th-ch**
This PR implements a raw version of PiP to resolve https://github.com/th-ch/youtube-music/issues/656 - first version re-uses the main window as having a new one requires way more work.

### CODE REVIEW

**Comment by Araxeus** [comment]
Doesn't really look like PiP to me, it looks like the native fullscreen mode (`f`) except its windowed

also it seems like pressing the "PiP" button cause the fullscreen mode to be windowed (but with the fullscreen ui)

so maybe this should just be an option (like "Windowed Fullscreen Mode") in the main menu? (Its more of a global toggle so I don't think it should be in the song menu)

There's Also an error/bug: ⚠️ 
trying to log in after having used "PiP" causes the app to freeze, console says
`TypeError: Cannot read properties of undefined (reading 'hide')
    at togglePiP (C:\Git\youtube-music\plugins\picture-in-picture\back.js:29:12)`

File: plugins/picture-in-picture/back.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		originalPosition = win.getPosition();
+		originalSize = win.getSize();
+
+		win.setFullScreenable(false);
+		await win.webContents.executeJavaScript(
+			// Go fullscreen
+			`document.querySelector(".fullscreen-button").click()`
+		);
+
+		app.dock.hide();
```
Comment (Araxeus) [inline plugins/picture-in-picture/back.js]:
```suggestion
		app.dock?.hide();
```

File: plugins/picture-in-picture/back.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		win.setVisibleOnAllWorkspaces(true, {
+			visibleOnFullScreen: true,
+		});
+		app.dock.show();
```
Comment (Araxeus) [inline plugins/picture-in-picture/back.js]:
```suggestion
		app.dock?.show();
```

File: plugins/picture-in-picture/back.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		win.setVisibleOnAllWorkspaces(false);
+		win.setAlwaysOnTop(false);
+	}
+
+	const [x, y] = isInPiPMode ? pipPosition : originalPosition;
+	const [w, h] = isInPiPMode ? pipSize : originalSize;
+	win.setPosition(x, y);
+	win.setSize(w, h);
+
+	win.setWindowButtonVisibility(!isInPiPMode);
```
Comment (Araxeus) [inline plugins/picture-in-picture/back.js]:
```suggestion
	win.setWindowButtonVisibility?.(!isInPiPMode);
```

**Review by Araxeus — CHANGES_REQUESTED** [review]

**Comment by Araxeus** [comment]
Ok sorry i'm just immediately writing as I'm thinking of stuff but ignore what I said about the size and name and all that

It just needs to behave correctly when entering fullscreen mode because right now thats buggy

also there should be a plugin option to set a hotkey for PiP

Good work 😄

**Comment by Araxeus** [comment]
2 more changes related to other plugins that should maybe come in follow up PR?
* precise-volume hud position needs to be tweaked if both pip AND in-app-menu are active
  (the volume hud is positioned inside the in-app-titlebar somehow)
* since in-app-menu is contained within the window, it shows only the topmost options in this mode, thus the most important menu options should maybe be moved up (View-exit/restart) (Options-AlwaysOnTop)

**Comment by Araxeus** [comment]
- [ ] right now its pretty annoying to exit the pip mode. there are 2 good solutions that I think could be both implemented:
  1. in `togglePiP` when PiP is activated ,register click event for t
   `$('.exit-fullscreen-button').addEvent -> onClick = togglePiP()` / `togglePictureInPicture()`
  and when its deactivated restore it
  2. like I wrote above, a custom keybind to toggle PiP

- [ ] on PiP start: if browserWindow is maximized, unmaximize it to avoid weird interaction

- [ ] some more css changes for more visibility in the playerbar:
  ```css
  ytmusic-player-bar.pip svg,
  ytmusic-player-bar.pip .time-info,
  ytmusic-player-bar.pip yt-formatted-string,
  ytmusic-player-bar.pip .yt-formatted-string {
	  filter: drop-shadow(2px 4px 6px black);
	  color: white !important;
	  fill: white !important;
  }
  ```
   - fix playerbar's artist/album string which instead of being `yt-formatted-string`  element they are links that have that as class.
   - fix `.timeInfo` which is a span that shows the current time (only when window is bigger than default PiP but still)
   - fix add `!important` to color to fix it not applying to artist name
   - fix `fill: white !important` mainly to fix `#button.dropdown-trigger` that didn't get filled by using `color

**Comment by th-ch** [comment]
Interesting feedback, thanks for the involvement! The PR is getting big and it gives a first version of the plugin (not perfect but good enough for a first iteration) so I'll merge it as is, and we can follow up with the remaining improvements (which make sense)! 🚀

**Comment by Araxeus** [comment]
>  I don't think that alwaysontop should be enabled by default since you can just enable it in the options
(maybe a plugin option to automatically enable alwaysontop but it shouldn't force this behavior)

^ do you agree with that?

should I open PR's about stuff I wrote since you closed this?

**Comment by th-ch** [comment]
definitely, making it a user choice is better than forcing it!

**Comment by Araxeus** [comment]
also im thinking maybe its better if the PiP button is inserted on the video overlay (like the "quality changer" button )
image
which would be way easier to access

(and then to exit that you could press the exit-fullscreen button i mentioned above / use the custom keybind)

**Pull request merged**